### DISCUSSION TITLE
Key mapping

### DISCUSSION THREAD

**Discussion by rageworx** [fire-eggs/YAIV#25] [Ideas]
Currently YAIV has strange key maps for me but it seems work like MacOS preview.
MacOS preview ( a window when space key on file navagator ) shows image/video as fitted in window client size, but it's not for detailed watching.
Maybe it depends on user's experience where it belong.
But I'm believe my suggestion may more friendly than default key maps ( or use minIni or tinyxml for change keymap )

How about this common platform dependency key schematics ( it is similar with HoneyView ) ?
- Arrow keys ( 4 directions )
    - basically it works as file seeking directions when image fitted inside window client.
    - but it works as image viewport scrolling when image is larger than window client region.
- Page up/down 
    - Previous/Next image file
-  Space 
    - Next image file
- Backspace/Delete(MacOS)
    - Previous image file
- Ctrl/CMD + Backspace/Delete(MacOS)
    - Remove current image ( delete file )
- Ctrl/CMD + C
    - Copy image to clipboard
- Ctrl+mouse wheel or +/-
    - zoom in/out
- Enter/Return
    - Fit to window client region.

## Additional references,
- Keymap exchanging of [Anystreaming](https://www.indiegogo.com/projects/anystreaming#/) client ( it was my startup project ) 
    - https://bitbucket.org/rageworx/astrm_w32/src/master/src/askeymap.cpp
    - https://bitbucket.org/rageworx/astrm_w32/src/master/src/askeyexchangetools.cpp
    - I was made hardware and embeded linux inside that device but it not released to the world...
- Hot key changing on Windows UI
    - It's FLTK screenshot program ( support multi languages by default LANGIDs with UTF-8 through tinyXML2 ) for Windows.
    - https://bitbucket.org/rageworx/flscrnshot/src/e36dbca0ec94d6409f5c596e2d4f07c994363471/src/winkeyhook.cpp#lines-98
    - This program using libturbojpeg to write JPEG in extramly fastest way.

**Comment by rageworx** [comment]
typo errors, sorry about my bad english ...

**Reply by fire-eggs** [reply]
Please don't apologize for your english - as a backward, monolingual American, I'm very impressed by anyone who has multiple languages!

**Comment by fire-eggs** [comment]
Good ideas here! The current key mappings are arbitrary except some are pulled from `feh`. The best long-term solution is, as you imply, to have different key-sets in external config files.

Have you ever created any means to let the user configure key settings? For [ImageGlass](https://github.com/d2phap/ImageGlass) [another project I collaborate on] I'm playing with a WinForms key-config control where the user types a key combo and it is shown, e.g. "CTRL+ALT+F5". I haven't even thought about the FLKT equivalent yet ...

**Reply by rageworx** [reply]
> Have you ever created any means to let the user configure key settings? For [ImageGlass](https://github.com/d2phap/ImageGlass) [another project I collaborate on] I'm playing with a WinForms key-config control where the user types a key combo and it is shown, e.g. "CTRL+ALT+F5". I haven't even thought about the FLKT equivalent yet ...

Yes I know your name on 2nd in 'Hall of fame' 😃
![image](https://user-images.githubusercontent.com/6948225/119601353-fb0d0780-be23-11eb-8af4-cbb3314fc86c.png)

And I'd checked before source was C# abd based on .NET framework.
I think C#(w/.NET) + [imageMagik ](https://imagemagick.org/index.php) [.NET version](https://github.com/dlemstra) may good choice to make a great viwer like imageGlass.

And double special keys like CTRL+ALT may implement to Fl_Widget, it is possible !
Here's sample what I made.

- ![image](https://user-images.githubusercontent.com/6948225/119610290-f4868c00-be33-11eb-8e41-da76ab216dd0.png)
- [fltk_keytest.zip](https://github.com/fire-eggs/YAIV/files/6544152/fltk_keytest.zip)

**Reply by fire-eggs** [reply]
Thanks for the sample, it motivated me to make a little progress! 

I'm attaching my modified version of the code which shows what I wanted for the "human friendly" string, as well as the encoding value which would be used to test against in a `handle()` function.

[main.cpp.zip](https://github.com/fire-eggs/YAIV/files/6572396/main.cpp.zip)

**Reply by rageworx** [reply]
> I'm attaching my modified version of the code which shows what I wanted for the "human friendly" string, as well as the encoding value which would be used to test against in a `handle()` function.

![image](https://user-images.githubusercontent.com/6948225/120259900-c7245d00-c2cf-11eb-8207-ec3bbf8afa7d.png)

Here I made something handle() mechanism as a callback pointer.
And any problem in shift ? ( you left 'irrelevant' in shift ... )

[20210531.zip](https://github.com/fire-eggs/YAIV/files/6573047/20210531.zip)

type make to test in your POSIX ( Linux or MSYS2 ), then run bin/test.
Hope it should be helpful for your next step.

plus, event calling by FL_KEYUP maybe better ... ?

**Reply by fire-eggs** [reply]
>And any problem in shift ?

I'll take another look - I think I quit too early playing with SHIFT.

In some apps, "SHIFT+CTRL+V" is distinct from "CTRL+V".

Here is a dilemma: should "CTRL+SHIFT+=" be shown as "CTRL++" or "CTRL+SHIFT+=" ? [American keyboard layout]. "CTRL+:" vs "CTRL+SHIFT+;". Etc.

**Reply by rageworx** [reply]
> Here is a dilemma: should "CTRL+SHIFT+=" be shown as "CTRL++" or "CTRL+SHIFT+=" ? [American keyboard layout]. "CTRL+:" vs "CTRL+SHIFT+;". Etc.

Stange result ... Korean 101 key implements US 101, and I'm using that ( my keyboards are all from US stores, no korean prints-only alphabets ) , just difference is right ALT to changing language Korean/English, and right Ctrl to HanJa(Kanji as japanese, but it's totatlly different)/some special charactors of Korean like ★. ⓐ, ⑽, ㅘ, ┎, ≪ . Korean keyboard layout actually same as US. ( Only Chinese or Japanese has big different variations, Only Korean keyboard don't need to extra keys for doing something more input )
What type of keyboards do you tested? I'm using mechanical keyboards ( n-key roll over ) and laptop default keyboards. Some keyboards cannot handle Ctrl+Alt+Shift+'some key' at once by reason of internal circuit limitation - so I'm using mechanical keyboards to prevent losing any key input.

You can test how many keys handled by your keyboard input by my FLTK keyboard test here.
https://bitbucket.org/rageworx_applications/rkkeyboarder/downloads/

![image](https://user-images.githubusercontent.com/6948225/120359461-a812e300-c342-11eb-9977-5ef5c6ede48b.png)

This is built for Windows, but you can build source for Linux or MacOS too.
And it belong to default FLTK keyboard testing source.
If your keyboard had some circuit limit on near Alt-Ctrl-Shift and ASDZXCQWE, this app also cannot display pressed key too. ( recommend to testing with N-key roll over something keyboard with USB or PS/2 interface )

**Reply by rageworx** [reply]
And default Korean keyboard layout you can see here, it's Lenovo thinkpad - UltraNav korean layout = US layout.
- https://rageworx.pe.kr/1756
( this is my Korean blog )

**Reply by fire-eggs** [reply]
This might be less about the keyboard [I'm using a mechanical keyboard] and more about how FLTK is doing key mapping.

The more I think about it, I'm not very concerned about "CTRL+:" vs "CTRL+SHIFT+;". Too much work, too many variables to be worth the effort.

**Reply by rageworx** [reply]
> The more I think about it, I'm not very concerned about "CTRL+:" vs "CTRL+SHIFT+;". Too much work, too many variables to be worth the effort.

There's maybe other way to detect right keys ( or emulate polling rate by pressed time and released like key-polling-scan ? )
I will try to find better way for platform-independency-widget.

**Comment by rageworx** [comment]
And C# ...

Some developers near me, they are love to using C#, No need to understand what GDI doing, easily design Visual Stuido and XAML ( or RESX ) on IDE. ( I was used Borland Delphi to make windows program quickly before about 10 years, maybe I believe it is similar approach )
But it is limited to experimental usage in company, because IL too easily reverse compiled, and it not be able to published with some security - I believe you also understand about [IL](https://en.wikipedia.org/wiki/Common_Intermediate_Language) as well and how weak at reverse engineering like by using [dnSpy](https://github.com/dnSpy/dnSpy).

But todays, maybe, .NET framework is widely used with MONO(Xamarin) for other platforms used for MS windows or Unity scripting as I know ... but still it had too big dependency for small systems like embedded linux on AARCH64.
I believe some people like me - using MacOS or Linux than Windows - may not like to start GUI project with C# by reason of many compatibility issues.

**Reply by fire-eggs** [reply]
I've done a fair amount of C# but until recently no one could program in it except on Windows. Even now, it seems too hard to make cross-platform GUI programs using C#. 

My main reason to switch to FLTK was for cross-platform GUI development, as I'm moving away from Windows to Linux as my main platform.

C# might have some advantages, especially when prototyping and experimenting, but I think you might just as well be using Java!

**Reply by rageworx** [reply]
> I think you might just as well be using Java!

No ... I'm not well be using JAVA ... I was been made JNI ( Java Native Interface ) for my RTOS platform ( name was DylOS, made for LG featurephones without Qulacomm BREW ) with C.
But JAVA was not my favor one ( until now ). More loved Object-Pascal (delphi).
The best - I think - is C and C++, and asm.

I used inline asm code ( for X86, at DOS era ) for making mode 13h ( MCGA 320x200, or 320x240, paletted 256 color ) games with Ad-Lib or Sound Blaster sound cards for directly accessing ISA ( 16bit BUS ).

I believe you have similar experiences.

**Reply by fire-eggs** [reply]
I don't care for Java myself. In comparison to C# though, with Java you can [supposedly] create a cross-platform app, with a GUI, and garbage collection, etc. So Java "seems" very similar to the capabilities of C#, _plus_ cross-platform GUI ...

Assembly language is in fact my weak spot. I did assembly language for Z80, 6502 and VAX and not much after that ...

**Reply by rageworx** [reply]
> Assembly language is in fact my weak spot. I did assembly language for Z80, 6502 and VAX and not much after that ...

I have experience of assembly programming for Z80, 6502, PIC at university lab ( there's no choice for C or C++ for performance at timing control as I know ), then continued for Renesas (Hitachi) and ATmel AVR series was so happy to I can make programming in C and C++. Sure I believe we know these kinds has no related things to consider multi-platform. 😃

Maybe ... JAVA is a good choice for platform-independancy base, but as we know .. there's hard to find GOOD-TO-USE or LOOKS-GOOD somethings. Plus. JAVA is too heavy and slow for some small systems, even there's issue on license with Oracle.

Anyway C and C++ will be the best language until end of human history, as like we saw Trinity accessing UNIX system V in Matrix movie. 😁

**Reply by rageworx** [reply]
>  So Java "seems" very similar to the capabilities of C#

C# and VB, those are similar with JAVA, right - I think so 😃
Only difference maybe container is EXE instead of JAR ( zip - deflate archive, most like Android APK ) that contains IL codes.

**Comment by rageworx** [comment]
# Key mapping issue on MacOS !
* MacOS has some different key combinitions,
* Super key = CMD key
* Alt = option key 
* Ctrl = Ctrl but cannot used by user application, specially ctrl + arrow keys are controlling desktop windows as like COMPIZ desktop window manager of Linux.
* Maybe CMD ( Super ) key instead on Linux...

**Reply by rageworx** [reply]
Tested for my forked repo, commited, refer to https://github.com/rageworx/YAIV/issues/10

**Reply by fire-eggs** [reply]
Changes integrated locally.
